keras_cv/layers/preprocessing/cut_mix.py

```diff
@@ -0,0 +1,142 @@
+
+
+class CutMix(layers.Layer):
+    """
+    CutMix implements the CutMix data augmentation technique as proposed in https://arxiv.org/abs/1905.04899.
+
+    Args:
+        num_classes: number of classes in the dataset.
+        alpha: alpha parameter for the sample distribution.
+        probability: probability to apply the CutMix augmentation.
+        label_smoothing: coefficient used in label smoothing.
+
+    Sample usage:
+    ```python
+    (x_train, y_train), (x_test, y_test) = keras.datasets.cifar10.load_data()
+    cutmix = CutMix(10)
+    augmented_data, updated_labels = cutmix(x_train, y_train)
+    ```
+    """
+    def __init__(
+        self, num_classes, alpha=0.8, probability=1.0, label_smoothing=0.0, **kwargs
+    ):
+        super(CutMix, self).__init__(*kwargs)
+        self.alpha = alpha
+        self.probability = probability
+        self.num_classes = num_classes
+        self.label_smoothing = label_smoothing
+
+    @staticmethod
+    def _sample_from_beta(alpha, beta, shape):
+        sample_alpha = tf.random.gamma(shape, 1.0, beta=alpha)
+        sample_beta = tf.random.gamma(shape, 1.0, beta=beta)
+        return sample_alpha / (sample_alpha + sample_beta)
+
+    def call(self, images, labels):
+        augment_cond = tf.less(
+            tf.random.uniform(shape=[], minval=0.0, maxval=1.0), self.probability
+        )
+        # pylint: disable=g-long-lambda
+        cutmix_augment = lambda: self._update_labels(*self._cutmix(images, labels))
+        no_augment = lambda: (images, self._smooth_labels(labels))
+        return tf.cond(augment_cond, cutmix_augment, no_augment)
+
+    def _cutmix(self, images, labels):
+        """Apply cutmix."""
+        lambda_sample = CutMix._sample_from_beta(
+            self.alpha, self.alpha, labels.shape
+        )
+
+        ratio = tf.math.sqrt(1 - lambda_sample)
+
+        batch_size = tf.shape(images)[0]
+        image_height, image_width = tf.shape(images)[1], tf.shape(images)[2]
+
+        cut_height = tf.cast(
+            ratio * tf.cast(image_height, dtype=tf.float32), dtype=tf.int32
+        )
+        cut_width = tf.cast(
+            ratio * tf.cast(image_height, dtype=tf.float32), dtype=tf.int32
+        )
+
+        random_center_height = tf.random.uniform(
+            shape=[batch_size], minval=0, maxval=image_height, dtype=tf.int32
+        )
+        random_center_width = tf.random.uniform(
+            shape=[batch_size], minval=0, maxval=image_width, dtype=tf.int32
+        )
+
+        bbox_area = cut_height * cut_width
+        lambda_sample = 1.0 - bbox_area / (image_height * image_width)
+        lambda_sample = tf.cast(lambda_sample, dtype=tf.float32)
+
+        images = tf.map_fn(
+            lambda x: _fill_rectangle(*x),
+            (
+                images,
+                random_center_width,
+                random_center_height,
+                cut_width // 2,
+                cut_height // 2,
+                tf.reverse(images, [0]),
```

Comment:
Any reason to use reversed order of images as the replacement? When there is only one image in the batch, does it mean it use itself as the fill?

Reply:
So in fill we simply use a tf.where() on a mask, giving it the `fill` and `image` Tensors.  By reversing we ensure that the image receives a different image than itself, without having to do any sort of expensive shuffle operation.

You are correct that a single image in a batch will use itself as the fill.

Perhaps this is really not even a valid use of the layer, maybe we should warn on the single image case?

Reply:
i would expect single image input is a normal use case, and the reverse order of the input is kind of building a hidden correlation for the image it will generate (it pair the input images in a deterministic way).

Can u use tf.random.shuffle instead?

Reply:
I've created an implementation with tf.random.shuffle.

For now this is ok, but it does yield some images to be paired with themselves, thus netting unaugmented images